File: scripts/gr-book-clusters.py
"""
Prepare GoodReads book clusters.
"""

from docopt import docopt

import numpy as np
import pandas as pd

from bookdata import script_log

_log = script_log('gr-book-clusters')

_log.info('reading book IDs')
books = pd.read_parquet('gr-book-ids.parquet', columns=['book_id', 'work_id'], use_nullable_dtypes=True)
# books['work_id'] = books['work_id'].astype('UInt32')
books.info()
books = books.set_index('book_id')
_log.info('reading ISBN IDs')
isbns = pd.read_parquet('book-isbn-ids.parquet')
isbns.info()
_log.info('reading clusters')
clusters = pd.read_parquet('../book-links/isbn-clusters.parquet')
clusters = clusters.set_index('isbn_id')

_log.info('merging ISBN info')
merged = isbns.join(books, on='book_id')
_log.info('merging cluster info')
merged = merged.join(clusters, on='isbn_id')

_log.info('deduplicating clusters')
ids = merged[['book_id', 'work_id', 'cluster']].drop_duplicates()

_log.info('saving book ID files')
ids.to_parquet('gr-book-link.parquet', index=False, compression='zstd')

File: src/bin/gr-book-clusters.rs
use std::fs::File;


use polars::prelude::*;
use bookdata::prelude::*;

use anyhow::Result;

/// Attach GoodReads books to clusers.
#[derive(StructOpt)]
#[structopt(name="gr-book-clusters")]
pub struct BookClusters {
  #[structopt(flatten)]
  common: CommonOpts,
}

fn main() -> Result<()> {
  let options = BookClusters::from_args();
  options.common.init()?;

  info!("reading book IDs");
  let idr = File::open("gr-book-ids.parquet")?;
  let idr = ParquetReader::new(idr);
  let id_df = idr.finish()?;

  info!("reading ISBN IDs");
  let r = File::open("book-isbn-ids.parquet")?;
  let r = ParquetReader::new(r);
  let isbn_df = r.finish()?;

  info!("reading clusters");
  let r = File::open("../book-links/isbn-clusters.parquet")?;
  let r = ParquetReader::new(r);
  let clus_df = r.finish()?;

  info!("merging ISBNs");
  let merged = id_df.left_join(&isbn_df, "book_id", "book_id")?;
  let merged = merged.left_join(&clus_df, "isbn_id", "isbn_id")?;

  info!("deduplicating clusters");
  let subset = merged.select(&["book_id", "work_id", "cluster"])?;
  let mut subset = subset.drop_duplicates(false, None)?;

  info!("writing results");
  let w = File::create("gr-book-link.parquet")?;
  let mut w = ParquetWriter::new(w);
  w.finish(&mut subset)?;

  Ok(())
}
